### PR TITLE
Add automatic merge config setup docs

### PR DESCRIPTION
**Pull request by dhiller**
/cc @fgimenez

### CODE REVIEW

File: docs/add-merge-automation-to-your-repository.md
Comment (jean-edouard) [inline docs/add-merge-automation-to-your-repository.md:8]: Missing dash after component name

File: docs/add-merge-automation-to-your-repository.md
Comment (jean-edouard) [inline docs/add-merge-automation-to-your-repository.md:17]: `s/some/Some/`

File: docs/add-merge-automation-to-your-repository.md
Comment (jean-edouard) [inline docs/add-merge-automation-to-your-repository.md:24]: Even if it has `needs-rebase`? Why do we retest on approve anyway?

File: docs/add-merge-automation-to-your-repository.md
Comment (jean-edouard) [inline docs/add-merge-automation-to-your-repository.md:31]: I've wondered about the details of that "eventually", if you know the specific timing that'd be interesting IMO.

File: docs/add-merge-automation-to-your-repository.md
Comment (xpivarc) [inline docs/add-merge-automation-to-your-repository.md:6]: https://github.com/kubernetes/test-infra/tree/master/prow/plugins/approve maybe?

File: docs/add-merge-automation-to-your-repository.md
Comment (jean-edouard) [inline docs/add-merge-automation-to-your-repository.md:26]: Could be good to add a note about `approve` not being removed, like:
Note: The `approve` label is not affected by PR modifications.

**Review by jean-edouard — COMMENTED** [review]
Very informative, thank you!
Minor comments in-line.

**Comment by jean-edouard** [comment]
Also, GitHub approval auto-adds `lgtm`, might be worth mentioning.

**Comment by jean-edouard** [comment]
Sorry for the spam, but there's a behavior I never quite understood with that system.
There's a PR that was approved, with all tests passing, and just needed an `lgtm` to be merged.
I added an lgtm, an now tide says `Pending — Not mergeable. Retesting 18 jobs.`, why?
(The PR in question https://github.com/kubevirt/kubevirt/pull/5944)

**Comment by fgimenez** [comment]
> Sorry for the spam, but there's a behavior I never quite understood with that system.
> There's a PR that was approved, with all tests passing, and just needed an `lgtm` to be merged.
> I added an lgtm, an now tide says `Pending — Not mergeable. Retesting 18 jobs.`, why?
> (The PR in question [kubevirt/kubevirt#5944](https://github.com/kubevirt/kubevirt/pull/5944))

Probably something got merged and the base branch changed, tide needs to make sure that the tests pass on your changes applied to the current state of the base branch. You can read more about tide here https://github.com/kubernetes/test-infra/blob/master/prow/cmd/tide/config.md#configuring-presubmit-jobs and here https://github.com/kubernetes/test-infra/blob/master/prow/cmd/tide/pr-authors.md

File: docs/add-merge-automation-to-your-repository.md
Comment (fgimenez) [reply docs/add-merge-automation-to-your-repository.md:24]: I think it refers to our `/retest` bot, it only sets the retest comment on approved PRs.

File: docs/add-merge-automation-to-your-repository.md
Comment (fgimenez) [inline docs/add-merge-automation-to-your-repository.md:3]: I think it would be good to distinguish here between actual Prow components (tide and branch-protector) and plugins (lgtm, approve, trigger), wdyt?

File: docs/add-merge-automation-to-your-repository.md
Comment (fgimenez) [inline docs/add-merge-automation-to-your-repository.md:10]: Maybe we could mention here also tide's `missingLabels`?

File: docs/add-merge-automation-to-your-repository.md
Comment (fgimenez) [reply docs/add-merge-automation-to-your-repository.md:31]: Agree, maybe a link to https://github.com/kubernetes/test-infra/blob/master/prow/cmd/tide/pr-authors.md could be useful.

**Comment by jean-edouard** [comment]
> > Sorry for the spam, but there's a behavior I never quite understood with that system.
> > There's a PR that was approved, with all tests passing, and just needed an `lgtm` to be merged.
> > I added an lgtm, an now tide says `Pending — Not mergeable. Retesting 18 jobs.`, why?
> > (The PR in question [kubevirt/kubevirt#5944](https://github.com/kubevirt/kubevirt/pull/5944))
> 
> Probably something got merged and the base branch changed, tide needs to make sure that the tests pass on your changes applied to the current state of the base branch. You can read more about tide here [https://github.com/kubernetes/test-infra/blob/master/prow/cmd/tide/config.md#configuring-presubmit-jobs](https://github.com/kubernetes/test-infra/blob/master/prow/cmd/tide/config.md?rgh-link-date=2021-07-08T14%3A10%3A43Z#configuring-presubmit-jobs) and here [https://github.com/kubernetes/test-infra/blob/master/prow/cmd/tide/pr-authors.md](https://github.com/kubernetes/test-infra/blob/master/prow/cmd/tide/pr-authors.md?rgh-link-date=2021-07-08T14%3A10%3A43Z)

Ah thanks, I wonder if that's configurable, because it seems highly unlikely that (non-conflicting) PRs would break all those tests we just spent hours running.
Having to rerun everything consumes a lot of time and resources, I would be surprised if fixing the occasional breakage isn't more efficient.
Maybe there could even be just 1 or 2 lanes that re-run in that scenario instead if all of them, further lowering breakage probability.

Anyway, just a thought, probably doesn't belong to this PR, sorry!

**Review by fgimenez — COMMENTED** [review]
Great to have this!

File: docs/add-merge-automation-to-your-repository.md
Comment (dhiller) [reply docs/add-merge-automation-to-your-repository.md:6]: Yeah, I just skipped that as there's no helpful README.md there. Maybe this one is more helpful, although I don't really like deep links to the code for a document: https://github.com/kubernetes/test-infra/blob/master/prow/plugins/approve/approve.go#L132

Also the lgtm README points towards a much more comprehensive documentation of the workflow.

> Also, GitHub approval auto-adds `lgtm`, might be worth mentioning.

Not in every case, but for our config you are right: https://github.com/kubevirt/project-infra/blob/main/github/ci/prow-deploy/kustom/base/configs/current/plugins/plugins.yaml#L463

Thanks, will add that!

**Comment by dhiller** [comment]
> Also, GitHub approval auto-adds `lgtm`, might be worth mentioning.

Good point, will add that. Please note that this is configured here: https://github.com/kubevirt/project-infra/blob/main/github/ci/prow-deploy/kustom/base/configs/current/plugins/plugins.yaml#L463

Thanks, will add that!

File: docs/add-merge-automation-to-your-repository.md
Comment (dhiller) [reply docs/add-merge-automation-to-your-repository.md:24]: TBH that's a very good point, maybe we should add that `needs-rebase` label at the periodic job definition: https://github.com/kubevirt/project-infra/blob/97ce8b6cc7bf8c66c58e02f47c1ce31e580c8181/github/ci/prow-deploy/files/jobs/kubevirt/project-infra/project-infra-periodics.yaml#L17

WDYT?

Comment (fgimenez) [reply docs/add-merge-automation-to-your-repository.md:24]: Sounds very good :+1:

**Comment by dhiller** [comment]
> > > Sorry for the spam, but there's a behavior I never quite understood with that system.
> > > There's a PR that was approved, with all tests passing, and just needed an `lgtm` to be merged.
> > > I added an lgtm, an now tide says `Pending — Not mergeable. Retesting 18 jobs.`, why?
> > > (The PR in question [kubevirt/kubevirt#5944](https://github.com/kubevirt/kubevirt/pull/5944))
> > 
> > 
> > Probably something got merged and the base branch changed, tide needs to make sure that the tests pass on your changes applied to the current state of the base branch. You can read more about tide here [https://github.com/kubernetes/test-infra/blob/master/prow/cmd/tide/config.md#configuring-presubmit-jobs](https://github.com/kubernetes/test-infra/blob/master/prow/cmd/tide/config.md?rgh-link-date=2021-07-08T14%3A10%3A43Z#configuring-presubmit-jobs) and here [https://github.com/kubernetes/test-infra/blob/master/prow/cmd/tide/pr-authors.md](https://github.com/kubernetes/test-infra/blob/master/prow/cmd/tide/pr-authors.md?rgh-link-date=2021-07-08T14%3A10%3A43Z)
> 
> Ah thanks, I wonder if that's configurable, because it seems highly unlikely that (non-conflicting) PRs would break all those tests we just spent hours running.
> Having to rerun everything consumes a lot of time and resources, I would be surprised if fixing the occasional breakage isn't more efficient.
> Maybe there could even be just 1 or 2 lanes that re-run in that scenario instead if all of them, further lowering breakage probability.
> 
> Anyway, just a thought, probably doesn't belong to this PR, sorry!

Tide also does try to batch test the pool of ready-to-merge PRs and then merge them all at the same time. (TBH on what conditions it tries that goes beyond my knowledge.) If the tests fail, IIUC it then tries to redo the testing for the PRs one at a time. 
@fgimenez please keep me honest here, but I think the eventual batch testing and merge of a set of PRs can be much more efficient especially when we have long running lanes.

File: docs/add-merge-automation-to-your-repository.md
Comment (dhiller) [reply docs/add-merge-automation-to-your-repository.md:3]: I didn't want to go into too much detail, but you are right, let's be clear here.

File: docs/add-merge-automation-to-your-repository.md
Comment (dhiller) [reply docs/add-merge-automation-to-your-repository.md:24]: Oh, I just saw it's already there: https://github.com/kubevirt/project-infra/blob/97ce8b6cc7bf8c66c58e02f47c1ce31e580c8181/github/ci/prow-deploy/files/jobs/kubevirt/project-infra/project-infra-periodics.yaml#L26

Comment (dhiller) [reply docs/add-merge-automation-to-your-repository.md:24]: So jed, you are right, as long as the need-rebase is present, no retesting by the bot will occur

**Comment by dhiller** [comment]
Folks, thanks for your thorough reviews, I've addressed the comments, PTAL!

**Comment by fgimenez** [comment]
> @fgimenez please keep me honest here, but I think the eventual batch testing and merge of a set of PRs can be much more efficient especially when we have long running lanes.

Exactly, batches allow tide to require fewer jobs run per PR. They have always priority over testing single PRs.

**Review by fgimenez — APPROVED** [review]
/approve

**Review by jean-edouard — APPROVED** [review]
/pony You forgot to document me!

**Comment by dhiller** [comment]
TIL about the [Safe Pony API](https://theponyapi.com/docs) :tada: :rofl:

**Comment by dhiller** [comment]
/pony princess celestia, what i learned today

File: docs/add-merge-automation-to-your-repository.md
Comment (stu-gott) [inline docs/add-merge-automation-to-your-repository.md]: [approve] ?

File: docs/add-merge-automation-to-your-repository.md
Comment (stu-gott) [inline docs/add-merge-automation-to-your-repository.md]: hyphen - after [branch-protection] ?

File: docs/add-merge-automation-to-your-repository.md
Comment (stu-gott) [inline docs/add-merge-automation-to-your-repository.md]: some presubmit jobs have `optional: false` set, which then requires their status to be successful

File: docs/add-merge-automation-to-your-repository.md
Comment (stu-gott) [inline docs/add-merge-automation-to-your-repository.md:32]: I don't understand what "none of the missing labels" means in this context. did you mean "and none of the labels are missing"?

Comment (stu-gott) [reply docs/add-merge-automation-to-your-repository.md:32]: also suggest "all required status results are successful"

**Review by stu-gott — COMMENTED** [review]
Looks good

**Comment by stu-gott** [comment]
/cc @mazzystr 

Over to you, sir. I believe you might be the sole approver entry in the OWNERS file.

**Pull request merged**
